syntax = "proto2";

/*
The IHDR chunk must appear FIRST. It contains:
   Width:              4 bytes
   Height:             4 bytes
   Bit depth:          1 byte
   Color type:         1 byte
   Compression method: 1 byte
   Filter method:      1 byte
   Interlace method:   1 byte
*/
message IHDR {
  required uint32 width  = 1;
  required uint32 height = 2;
  required uint32 other1 = 3;  // all 4 bytes are used
  required uint32 other2 = 4;  // only 1 byte used!
}

// bytes corresponds to string in C++
message PLTE {
  required bytes data = 1;
}

message IDAT {
  required bytes data = 1;
}

message iCCP  {
  required bytes data = 2;
}

message OtherChunk {
  oneof type {
    uint32 known_type = 1;
    uint32 unknown_type = 2;
  }
  required bytes data = 3;
}

message PngChunk {
  oneof chunk {
    PLTE plte = 1;
    IDAT idat = 2;
    iCCP iccp = 3;
    OtherChunk other_chunk = 10000;
  }
}

// The first 8 bytes of a PNG file is the PNG Signature (constant)
message PngProto {
  required IHDR ihdr = 1;
  repeated PngChunk chunks = 2;
}

